### PR TITLE
Mongodb

### PR DESCRIPTION
**Pull request by NevarDima**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -20,6 +20,7 @@ repositories {
 
 dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-webflux'
+	implementation 'org.springframework.boot:spring-boot-starter-data-mongodb-reactive:2.7.1'
 	compileOnly 'org.projectlombok:lombok'
```
Comment (skalashynski) [inline build.gradle]: move please version number to section `versions`

File: src/main/java/com/qashqade/documentservice/handler/Handler.java [hunk 2]
```diff
@@ -1,41 +1,36 @@
...
-    private final MessageService messageService;
+    private final WebMessageService webMessageService;
 
     public Mono<ServerResponse> list(ServerRequest request) {
         var limit = Integer.parseInt(request.queryParam("limit").orElse("0"));
 
-        Flux<Message> publisher = messageService.allMessages(limit);
-        publisher.map(m -> m.getId()).subscribe(id -> System.out.print(id+","));
-        System.out.println("\n--------------\n--------------\n");
+        webMessageService.allMessages(limit).doOnComplete(() -> System.out.println("WebMessageService done!"));
```
Comment (skalashynski) [inline src/main/java/com/qashqade/documentservice/handler/Handler.java]: It's not clear. Type: "200 messages were successfully saved to mongo db."

File: src/main/java/com/qashqade/documentservice/handler/Handler.java [hunk 3]
```diff
@@ -1,41 +1,36 @@
...
 
         return ServerResponse
             .ok()
-            .contentType(MediaType.APPLICATION_JSON)
-            .body(BodyInserters.fromValue(limit +" trxs"));
+            .contentType(MediaType.APPLICATION_NDJSON)
+            .body(BodyInserters.fromValue(limit +" trxs done"));
```
Comment (skalashynski) [inline src/main/java/com/qashqade/documentservice/handler/Handler.java]: It's not clear 'done'

File: src/main/java/com/qashqade/documentservice/service/MongoMessageService.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    public Mono<Message> getById(Long id){
+        return mongoRepository.findById(id);
+    }
+
+    public Mono<Message> save (Mono<Message> messageMono){
+        long currentTime = System.currentTimeMillis();
+        return messageMono.flatMap(mongoRepository::insert)
+            .doOnSuccess(m -> System.out.println("id " + m.getId() + " message saved! " + (System.currentTimeMillis() - currentTime)/1000.0))
+            .doOnError(e -> System.out.println("Message can't be saved ().()"+e.getMessage()));
+    }
```
Comment (skalashynski) [inline src/main/java/com/qashqade/documentservice/service/MongoMessageService.java]: Please add message id to log

File: src/main/java/com/qashqade/documentservice/service/WebMessageService.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    public Flux<Message> allMessages(int limit){
+        long currentTimeToGet = System.currentTimeMillis();
+        String queryParam = "";
+        if (limit>0) queryParam = "?limit="+limit;
+        Flux<Message> messageFlux = webClient.get()
+            .uri("/messages" + queryParam)
+            .retrieve()
+            .bodyToFlux(Message.class)
+            .doOnComplete(() -> System.out.println("documentservice " + limit + " Yes! " + (System.currentTimeMillis() - currentTimeToGet)/1000.0))
+            .doOnError(e -> System.out.println("documentservice " + limit + "No ;(. " + e.getMessage()));
```
Comment (skalashynski) [inline src/main/java/com/qashqade/documentservice/service/WebMessageService.java]: Incorrect log messages.

**Review by skalashynski — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 1 above)
Comment (NevarDima) [reply build.gradle]: So we have plugin section with line "id 'org.springframework.boot' version '2.7.1'", we don't need to specify the version in the implementation

File: src/main/java/com/qashqade/documentservice/handler/Handler.java (hunk 2 above)
Comment (NevarDima) [reply src/main/java/com/qashqade/documentservice/handler/Handler.java]: Resolved and done console output: "200 messages were successfully saved to mongo db."

File: src/main/java/com/qashqade/documentservice/handler/Handler.java (hunk 3 above)
Comment (NevarDima) [reply src/main/java/com/qashqade/documentservice/handler/Handler.java]: Resolved and done console output: "WebMessageService has got " + limit + " transactions" (where limit is num of messages from request.queryParam)

File: src/main/java/com/qashqade/documentservice/service/MongoMessageService.java (hunk 4 above)
Comment (NevarDima) [reply src/main/java/com/qashqade/documentservice/service/MongoMessageService.java]: resolved and added id to console output: "Message id " + id[0] + " can't be saved ().()" + e.getMessage() (where id[0] is the id of message)

File: src/main/java/com/qashqade/documentservice/service/WebMessageService.java (hunk 5 above)
Comment (NevarDima) [reply src/main/java/com/qashqade/documentservice/service/WebMessageService.java]: Resolved and rewritten incorrect messages to more informative ones

File: src/main/java/com/qashqade/documentservice/handler/Handler.java (hunk 2 above)
Comment (skalashynski) [reply src/main/java/com/qashqade/documentservice/handler/Handler.java]: I meant, no exactly 200. Update to `limit` param

File: build.gradle (hunk 1 above)
Comment (NevarDima) [reply build.gradle]: Fixed.

File: src/main/java/com/qashqade/documentservice/service/DataFetchService.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        String queryParam = "";
+        if (limit > 0) queryParam = "?limit=" + limit;
+        Flux<Transaction> transactionFlux = webClient.get()
+            .uri("/transactions" + queryParam)
+            .retrieve()
+            .bodyToFlux(Transaction.class)
+            .log("TransactionService.allTransactions", Level.INFO, SignalType.ON_ERROR)
+            .doOnComplete(() -> log.debug("{} transactions have been got in {} seconds!", limit,
+                (System.currentTimeMillis() - currentTime) / 1000.0))
+            .doOnError(e -> log.error("{} can't be obtained!", limit, e));
```
Comment (skalashynski) [inline src/main/java/com/qashqade/documentservice/service/DataFetchService.java]: Change log message here

**Review by skalashynski — APPROVED** [review]

File: src/main/java/com/qashqade/documentservice/service/DataFetchService.java (hunk 6 above)
Comment (NevarDima) [reply src/main/java/com/qashqade/documentservice/service/DataFetchService.java]: Fixed and rewritten with log output "can't be obtained from repository module!"

**Pull request merged**